mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala

```diff
@@ -0,0 +1,238 @@
+ */
+class IsotonicRegression
+  extends Serializable {
+
+  /**
+   * Run algorithm to obtain isotonic regression model
+   *
+   * @param input (label, feature, weight)
+   * @param isotonic isotonic (increasing) or antitonic (decreasing) sequence
+   * @return isotonic regression model
+   */
+  def run(
+      input: RDD[(Double, Double, Double)],
+      isotonic: Boolean = true): IsotonicRegressionModel = {
+    createModel(
+      parallelPoolAdjacentViolators(input, isotonic),
+      isotonic)
+  }
+
+  /**
+   * Creates isotonic regression model with given parameters
+   *
+   * @param predictions labels estimated using isotonic regression algorithm.
+   *                    Used for predictions on new data points.
+   * @param isotonic isotonic (increasing) or antitonic (decreasing) sequence
+   * @return isotonic regression model
+   */
+  protected def createModel(
+      predictions: Array[(Double, Double, Double)],
+      isotonic: Boolean): IsotonicRegressionModel = {
+
+    val labels = predictions.map(_._1)
+    val features = predictions.map(_._2)
+
+    new IsotonicRegressionModel(features, labels)
+  }
+
+  /**
+   * Performs a pool adjacent violators algorithm (PAVA)
+   * Uses approach with single processing of data where violators
+   * in previously processed data created by pooling are fixed immediatelly.
+   * Uses optimization of discovering monotonicity violating sequences (blocks)
+   * Method in situ mutates input array
+   *
+   * @param in input data
+   * @param isotonic asc or desc
+   * @return result
+   */
+  private def poolAdjacentViolators(
+      in: Array[(Double, Double, Double)],
+      isotonic: Boolean): Array[(Double, Double, Double)] = {
+
+    // Pools sub array within given bounds assigning weighted average value to all elements
+    def pool(in: Array[(Double, Double, Double)], start: Int, end: Int): Unit = {
+      val poolSubArray = in.slice(start, end + 1)
+
+      val weightedSum = poolSubArray.map(lp => lp._1 * lp._3).sum
+      val weight = poolSubArray.map(_._3).sum
+
+      var i = start
+      while (i <= end) {
+        in(i) = (weightedSum / weight, in(i)._2, in(i)._3)
+        i = i + 1
+      }
+    }
+
+    val monotonicityConstraintHolds: (Double, Double) => Boolean =
+      (x, y) => if (isotonic) x <= y else x >= y
+
+    var i = 0
+    while (i < in.length) {
+      var j = i
+
+      // Find monotonicity violating sequence, if any
+      while (j < in.length - 1 && !monotonicityConstraintHolds(in(j)._1, in(j + 1)._1)) {
+        j = j + 1
+      }
+
+      // If monotonicity was not violated, move to next data point
+      if (i == j) {
+        i = i + 1
+      } else {
+        // Otherwise pool the violating sequence
+        // And check if pooling caused monotonicity violation in previously processed points
+        while (i >= 0 && !monotonicityConstraintHolds(in(i)._1, in(i + 1)._1)) {
+          pool(in, i, j)
```

Comment:
Instead of pooling `in` in-place, I recommend creating a buffer for new segments for data compression.